File: src/domain/garment-filtering.test.ts
import { describe, expect, it } from 'vitest';
import { filterGarments, sortGarments } from './garment-filtering';
import type { Garment } from '../types';

const garment = (overrides: Partial<Garment>): Garment => ({
  id: 'g',
  image_uri: 'a.jpg',
  image_uri_nobg: null,
  image_uris: ['a.jpg'],
  image_uris_nobg: [],
  category: 'tops',
  subcategory: 'T-Shirt',
  subcategories: ['T-Shirt'],
  tags: [],
  brand: null,
  color_primary: '#000000',
  color_secondary: null,
  color_palette: ['#000000'],
  size: 'M',
  purchase_date: null,
  is_available: true,
  unavailable_date: null,
  created_at: '2026-01-01T00:00:00.000Z',
  updated_at: '2026-01-01T00:00:00.000Z',
  ...overrides,
});

describe('filterGarments', () => {
  it('keeps everything when nothing is asked for', () => {
    const all = [garment({ id: 'a' }), garment({ id: 'b' })];
    expect(filterGarments(all).map(g => g.id)).toEqual(['a', 'b']);
  });

  it('matches a subcategory in the list or the singular column', () => {
    const inList = garment({ id: 'list', subcategories: ['Hoodie'], subcategory: null });
    const singular = garment({ id: 'single', subcategories: [], subcategory: 'Hoodie' });
    const other = garment({ id: 'other', subcategories: ['Polo'], subcategory: 'Polo' });

    expect(filterGarments([inList, singular, other], { subcategory: 'Hoodie' }).map(g => g.id))
      .toEqual(['list', 'single']);
  });

  it('matches a season tag case-insensitively', () => {
    const tagged = garment({ id: 'tagged', tags: ['Summer'] });
    const untagged = garment({ id: 'untagged', tags: [] });

    expect(filterGarments([tagged, untagged], { season: 'summer' }).map(g => g.id))
      .toEqual(['tagged']);
  });

  it('derives occasion from the garment type rather than its tags', () => {
    // Nothing is tagged 'work'; a Blouse is work-appropriate by type.
    const blouse = garment({ id: 'blouse', subcategories: ['Blouse'] });
    const tee = garment({ id: 'tee', subcategories: ['T-Shirt'] });

    expect(filterGarments([blouse, tee], { occasion: 'work' }).map(g => g.id)).toEqual(['blouse']);
  });

  it('matches brand and size as case-insensitive substrings', () => {
    const uniqlo = garment({ id: 'uniqlo', brand: 'Uniqlo', size: 'M' });
    const nike = garment({ id: 'nike', brand: 'Nike', size: 'XL' });

    expect(filterGarments([uniqlo, nike], { brand: '  uniQ ' }).map(g => g.id)).toEqual(['uniqlo']);
    expect(filterGarments([uniqlo, nike], { size: 'x' }).map(g => g.id)).toEqual(['nike']);
  });

  it('matches a colour anywhere in the palette, not just the primary', () => {
    const secondary = garment({
      id: 'secondary',
      color_primary: '#000000',
      color_palette: ['#000000', '#CC0000'],
    });
    const neither = garment({ id: 'neither', color_palette: ['#FFFFFF'] });

    expect(filterGarments([secondary, neither], { color: '#CC0000' }).map(g => g.id))
      .toEqual(['secondary']);
  });

  it('applies every filter given, not just the first', () => {
    const match = garment({ id: 'match', brand: 'Uniqlo', tags: ['summer'] });
    const wrongBrand = garment({ id: 'wrong', brand: 'Nike', tags: ['summer'] });
    const wrongSeason = garment({ id: 'season', brand: 'Uniqlo', tags: ['winter'] });

    expect(
      filterGarments([match, wrongBrand, wrongSeason], { brand: 'Uniqlo', season: 'summer' })
        .map(g => g.id)
    ).toEqual(['match']);
  });
});

describe('sortGarments', () => {
  it('puts the newest first by default', () => {
    const items = [
      garment({ id: 'old', created_at: '2026-01-01' }),
      garment({ id: 'new', created_at: '2026-06-01' }),
      garment({ id: 'mid', created_at: '2026-03-01' }),
    ];

    expect(sortGarments(items).map(g => g.id)).toEqual(['new', 'mid', 'old']);
    expect(sortGarments(items, 'oldest').map(g => g.id)).toEqual(['old', 'mid', 'new']);
  });

  it('does not reorder its input', () => {
    const items = [garment({ id: 'a', created_at: '2026-01-01' }), garment({ id: 'b', created_at: '2026-06-01' })];
    sortGarments(items);
    expect(items.map(g => g.id)).toEqual(['a', 'b']);
  });

  it('survives a garment with no timestamp', () => {
    // created_at is declared non-null but really can be null on an install
    // upgraded through the ALTER path. Dereferencing it threw for any wardrobe
    // big enough for the null to reach the right-hand side of a comparison, and
    // the list hook swallowed that -- so the screen showed an empty wardrobe.
    const items = Array.from({ length: 12 }, (_, i) => garment({
      id: `g${i}`,
      created_at: (i === 5 ? null : `2026-01-${String(i + 1).padStart(2, '0')}`) as string,
    }));

    const sorted = sortGarments(items);

    expect(sorted).toHaveLength(12);
    // An absent timestamp sorts as the earliest possible, so it goes last here
    // rather than vanishing.
    expect(sorted[sorted.length - 1].id).toBe('g5');
    expect(sortGarments(items, 'oldest')[0].id).toBe('g5');
  });
});

File: presentation/src/test/kotlin/com/wardrobapp/presentation/GarmentFilteringTest.kt
package com.wardrobapp.presentation

import com.wardrobapp.data.normalizeGarmentRow
import com.wardrobapp.domain.Occasion
import com.wardrobapp.domain.Season
import kotlin.test.Test
import kotlin.test.assertEquals

/**
 * Narrowing and ordering the wardrobe list.
 *
 * Filters combine, so the interesting failures are the ones where a filter is
 * ignored rather than the ones where it is wrong: a screen showing more than it
 * says it is showing. The ordering has a nastier failure behind it -- an install
 * upgraded through the ALTER path can have no `created_at` at all, and the
 * TypeScript this was ported from dereferenced it and threw, which the list hook
 * swallowed into an *empty wardrobe*.
 */
class GarmentFilteringTest {

    private fun garment(
        id: String,
        subcategories: List<String> = listOf("T-Shirt"),
        tags: List<String> = emptyList(),
        brand: String? = null,
        size: String? = null,
        palette: List<String> = listOf("#1F3A93"),
        createdAt: String? = "2026-01-01T00:00:00.000Z",
        category: String = "tops",
    ) = normalizeGarmentRow(
        mapOf(
            "id" to id,
            "image_uri" to "$id.jpg",
            "category" to category,
            "subcategories" to subcategories.joinToString(",", "[\"", "\"]"),
            "tags" to (if (tags.isEmpty()) "[]" else tags.joinToString("\",\"", "[\"", "\"]")),
            "brand" to brand,
            "size" to size,
            "color_primary" to palette.first(),
            "color_palette" to palette.joinToString("\",\"", "[\"", "\"]"),
            "created_at" to createdAt,
        ),
        "",
    )

    private val wardrobe = listOf(
        garment("shirt", listOf("Shirt"), tags = listOf("winter"), brand = "Uniqlo", size = "M"),
        garment("tee", listOf("T-Shirt"), tags = listOf("summer"), brand = "Arket", size = "L"),
        garment("blouse", listOf("Blouse"), brand = "arket", size = "m", palette = listOf("#C0392B")),
    )

    @Test
    fun `no filter is not a filter`() {
        assertEquals(wardrobe.map { it.id }, wardrobe.filterBy(GarmentFilter()).map { it.id })
    }

    @Test
    fun `each filter narrows by the thing it names`() {
        assertEquals(listOf("tee"), wardrobe.filterBy(GarmentFilter(subcategory = "T-Shirt")).map { it.id })
        assertEquals(listOf("shirt"), wardrobe.filterBy(GarmentFilter(season = Season.WINTER)).map { it.id })
        assertEquals(listOf("blouse"), wardrobe.filterBy(GarmentFilter(color = "#C0392B")).map { it.id })
    }

    @Test
    fun `an occasion is what the garment is, not what it was tagged`() {
        // Nothing here carries an occasion tag; a blouse is formal because of what
        // it is.
        assertEquals(listOf("blouse"), wardrobe.filterBy(GarmentFilter(occasion = Occasion.FORMAL)).map { it.id })
    }

    @Test
    fun `brand and size match loosely, because they are typed by hand`() {
        // Case and partial words, so "ark" finds Arket however it was capitalised.
        assertEquals(listOf("tee", "blouse"), wardrobe.filterBy(GarmentFilter(brand = "ark")).map { it.id })
        assertEquals(listOf("shirt", "blouse"), wardrobe.filterBy(GarmentFilter(size = "M")).map { it.id })
        // A blank needle is not a filter at all rather than matching nothing.
        assertEquals(3, wardrobe.filterBy(GarmentFilter(brand = "  ")).size)
    }

    @Test
    fun `filters combine`() {
        assertEquals(
            listOf("blouse"),
            wardrobe.filterBy(GarmentFilter(brand = "arket", size = "m")).map { it.id },
        )
        // And a combination nothing satisfies returns nothing rather than the
        // last filter's answer.
        assertEquals(
            emptyList(),
            wardrobe.filterBy(GarmentFilter(brand = "uniqlo", size = "L")).map { it.id },
        )
    }

    @Test
    fun `newest first, oldest last, and back again`() {
        val list = listOf(
            garment("old", createdAt = "2025-01-01T00:00:00.000Z"),
            garment("new", createdAt = "2026-06-01T00:00:00.000Z"),
            garment("middle", createdAt = "2025-09-01T00:00:00.000Z"),
        )

        assertEquals(listOf("new", "middle", "old"), list.orderedBy(GarmentSort.NEWEST).map { it.id })
        assertEquals(listOf("old", "middle", "new"), list.orderedBy(GarmentSort.OLDEST).map { it.id })
        // Newest is the default, because that is what the screen opens on.
        assertEquals(list.orderedBy(GarmentSort.NEWEST).map { it.id }, list.orderedBy().map { it.id })
    }

    @Test
    fun `a garment with no timestamp sorts last rather than disappearing`() {
        // The failure this replaces was not a wrong order: it was an exception
        // that emptied the whole screen.
        val list = listOf(
            garment("dated", createdAt = "2026-01-01T00:00:00.000Z"),
            garment("undated", createdAt = null),
        )

        assertEquals(listOf("dated", "undated"), list.orderedBy(GarmentSort.NEWEST).map { it.id })
        assertEquals(listOf("undated", "dated"), list.orderedBy(GarmentSort.OLDEST).map { it.id })
    }
}
